localstack/aws/protocol/parser.py

```diff
@@ -145,49 +150,42 @@ def parse(self, request: HttpRequest) -> Tuple[OperationModel, Any]:
         if location is not None:
             if location == "header":
-                # TODO implement proper parsing from the header field
-                # https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html#httpheader-trait
-                # Attention: This differs from the other protocols!
-                # headers = request.headers
-                # location_name = shape.serialization.get("locationName")
-                payload = ""
-                raise NotImplementedError
+                header_name = shape.serialization.get("name")
+                if header_name in request.headers:
+                    payload = self._parse_shape(
+                        request, shape, request.headers[header_name], path_regex
+                    )
             elif location == "headers":
-                # TODO implement proper parsing from the header fields (prefix)
-                # https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html#httpprefixheaders-trait
-                # Attention: This differs from the other protocols!
-                # headers = request.headers
-                # location_name = shape.serialization.get("locationName")
-                payload = ""
-                raise NotImplementedError
+                payload = self._parse_header_map(shape, request.headers)
             elif location == "querystring":
-                # TODO implement proper parsing from the query string
-                # https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html#httpquery-trait
-                # Attention: This differs from the other protocols, even the Query protocol!
-                # body = request.text
-                # location_name = shape.serialization.get("locationName")
-                payload = ""
-                raise NotImplementedError
+                query_name = shape.serialization.get("name")
+                parsed_query = parse_qs(to_str(request.query_string))
+                # If the query param is in the parsed dict, take the first element of the value
+                # (the parsed query params are always lists, even if only a single value).
+                payload = (
+                    next(iter(parsed_query[query_name]), None)
+                    if query_name in parsed_query
+                    else None
+                )
             elif location == "uri":
-                # TODO implement proper parsing from the URI path
-                # https://awslabs.github.io/smithy/1.0/spec/core/http-traits.html#httplabel-trait
-                # Attention: This differs from the other protocols, even the Query protocol!
-                # path = request.path
-                # location_name = shape.serialization.get("locationName")
-                payload = ""
-                raise NotImplementedError
+                regex_group_name = shape.serialization.get("name")
+                match = path_regex.match(request.path)
```

Comment:
i see we are handing down this `path_regex` attribute through the entire stack, but only need it at this one location. i'm a bit fuzzy on the details in the parser here, but i was just wondering whether we can (for the next iteration) think about abstracting some context object we can hand down instead, which contains all the information necessary for parsing? just food for thought.

Reply:
Definitely. If we end up handling down another parameter, we should introduce some kind of context.
But since the spec features are somewhat limited (and we implement the vast majority of them at this point), I couldn't think of anything else that might come up later.